docs/keyed_services.md

```diff
@@ -0,0 +1,184 @@
+BrowserContextKeyedServices per
+[profile-architecture](https://www.chromium.org/developers/design-documents/profile-architecture/).
+What this means is that all associated preferences and other locally stored
+state will be per-profile. A service may sometimes be shared between the regular
+and OTR profile (check with sec-team) and it may also be ununavailable for
+certain profile types.
+`ProfileKeyedServiceFactory/ProfileKeyedServiceFactoryIOS`
+[profile_keyed_service_factory.md](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/profiles/profile_keyed_service_factory.md;bpv=0)
+with `ProfileSelections::Builder` is the preferred method to use for determining
+which (if any) profile should be used for the service.
+
+
+keyed-service-docs
+```cpp
+MyServiceFactory::MyServiceFactory()
+    : ProfileKeyedServiceFactory(
+          "MyService",
+          ProfileSelections::Builder()
+              .WithRegular(ProfileSelection::kOriginalOnly)
+              .Build()) { ... }
+```
+
+If you have more complicated logic then use
+`BrowserContextKeyedServiceFactory::GetBrowserContextToUse` and return nullptr
+if the service is not available for the given `BrowserContext`.
+
+```cpp
+content::BrowserContext* MyServiceFactory::GetBrowserContextToUse(
+    content::BrowserContext* context) const {
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsRegularProfile() &&
+         base::FeatureList::IsEnabled(features::kMyFeature);
+}
+```
+
+Do not explicitly return null from other methods like
+`BuildServiceInstanceForBrowserContext`, `GetForProfile`, etc... If you return
+nullptr from `GetBrowserContextToUse` or configure `ProfileSelections`, the
+service will already be null. Avoid using helper methods like
+`IsAllowedForProfile` to see if a service is available for a particular profile
+as these are often used incorrectly/inconsistently. The preferred method is just
+a null check on the service itself. Prefer dependency injection
+[structure-modularity](https://chromium.googlesource.com/chromium/src/+/main/docs/chrome_browser_design_principles.md#structure_modularity)
+and pass in the services you require instead of calling the factory methods
+internally. This reduces dependencies on the factories and makes unit
+testing/mocking simpler.
+
+#### Keyed Services and Mojo
+
+Avoid adding wrapper classes around keyed services to implement mojo interfaces.
+It's usually better to implement them directly in the service. Factories should
+implement `GetRemoteForProfile` to return a `mojo::PendingRemote` and
+`BindRemoteForProfile` when you are passing in a `mojo::PendingReceiver`. The
+keyed service should implement `MakeRemote()` to return the
+`mojo::PendingRemote`. Please always ensure that you check to make sure the
+service is not null before calling `MakeRemote()` and return an empty pending
+remote if it is `mojo::PendingRemote<mojom::FakeService>()`.
```

Comment:
`mojo::PendingRemote<mojom::FakeService>()` this reads as if the service should support a "Fake" version of it. I don't think you mean it here.

Let's use `ExampleService` instead of `FakeService` here and in other examples.

Reply:
Yeah I don't know why I was struggling to come up with something at the time, but Example is much better